### PR TITLE
Cache Plugin improvements

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
### Description

After adding some options added for URI exclusions by @OctavianC in the cache plugin (https://github.com/joomla/joomla-cms/pull/7767), this PR adds some other improvements:
- Checks SEF and non SEF URI
- Since it also checks the non sef URI now you can exclude all pages of a
  component (ex: com_users)
- No need to backslash the slashes
- Do not cache or load from cache if site is offline
- Some other minor code improvements
- Show debugger even in cached pages
### How to test

With a fresh Joomla install with sample items, turn on cache plugin and exclude some SEF or Non-SEF page URI in "Exclude Pages" field, for instance:
- com_contact
- option=com_content
- /login
- /article-categories-view
1. Check if pages is being cached or not.
2. Put the site offline and see that it does not generates or load from cache.
### B/C

Should be ok.
### Notes

Since i'm not expert in Joomla code, if there are betters ways to do this, improvements are welcome.

### CODE REVIEW

**Comment by andrepereiradasilva** [comment]
First time travis free in one go!

File: administrator/language/en-GB/en-GB.plg_system_cache.ini [hunk 1]
```diff
@@ -5,7 +5,7 @@
 
 PLG_CACHE_FIELD_BROWSERCACHE_DESC="If yes, use mechanism for storing page cache in the browser."
 PLG_CACHE_FIELD_BROWSERCACHE_LABEL="Use Browser Caching"
-PLG_CACHE_FIELD_EXCLUDE_DESC="Specify which URLs you want to exclude from caching, each on a separate line. Regular expressions are supported, eg. <br /><strong>about\-[a-z]+</strong> - will exclude all URLs that contain 'about-', for example 'about-us', 'about-me', 'about-joomla' etc.<br /><strong>\/component\/users\/</strong> - will exclude all URLs that contain /component/users/"
+PLG_CACHE_FIELD_EXCLUDE_DESC="Specify which URLs you want to exclude from caching, each on a separate line. Regular expressions are supported, eg. <br /><strong>about\-[a-z]+</strong> - will exclude all URLs that contain 'about-', for example 'about-us', 'about-me', 'about-joomla' etc.<br /><strong>/component/users/</strong> - will exclude all URLs that contain /component/users/.<br /><strong>com_users</strong> - will exclude users all component pages"
 PLG_CACHE_FIELD_EXCLUDE_LABEL="Exclude URLs"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_cache.ini]: Should the ending be
Will exclude all Users component pages.

**Comment by andrepereiradasilva** [comment]
@brianteeman Thanks. Done

**Comment by OctavianC** [comment]
I haven't tested the code, but there are two things I've noticed:
- I'm not really fond of the % regex delimiter. PHP examples use the forward slash and if you search on the web, almost everyone uses it as well.
- This expression `%<input type="hidden" name="[A-Za-z0-9]+" value="1"\s+/>%` will incorrectly replace these as well:

`<input type="hidden" name="agree" value="1" />`

`<input type="hidden" name="id" value="1" />`

`<input type="hidden" name="catid" value="1" />`

Basically all hidden fields that contain numbers and characters (that's a big range) will get replaced by the plugin and break functionality for the respective extensions.

**Comment by andrepereiradasilva** [comment]
> This expression %<input type="hidden" name="[A-Za-z0-9]+" value="1"\s+/>% will incorrectly replace these as well: [...] Basically all hidden fields that contain numbers and characters (that's a big range) will get replaced by the plugin and break functionality for the respective extensions.
The problem is when you cache a page with a form token, that form token will be cached too. So it will be served to all users. This code was to try to correct that.
And yes, you are right, my mistake. Any suggestions how to do this without breaking functionality.

> I'm not really fond of the % regex delimiter. PHP examples use the forward slash and if you search on the web, almost everyone uses it as well.
In PHP preg expressions you can use any character as delimiter. Since URI use a lot of slashes i thing is more suitable for this case using another character, that is why i used `%` instead of `/` as regex delimiter.

**Comment by OctavianC** [comment]
I'm not arguing about the purpose of the expression, I understand what it's trying to do, I was just saying that it doesn't work well.

**Comment by andrepereiradasilva** [comment]
Ok. I will see if i can get some way to replace the token without breaking something.
If not i will remove it from the code.

**Comment by OctavianC** [comment]
Looking at the code it seems `JCache::getWorkarounds()` already does what you're looking for, although I'm not sure where it's used. See 10b1f572d019dd53582c3af01fce0c2b4dd43239

**Comment by andrepereiradasilva** [comment]
> Looking at the code it seems JCache::getWorkarounds() already does what you're looking for, although I'm not sure where it's used. See 10b1f57

I didn't saw that. Will test.

**Comment by andrepereiradasilva** [comment]
Ok. done and some minor otimizations

**Comment by andrepereiradasilva** [comment]
@OctavianC can you check?

**Comment by OctavianC** [comment]
Sorry for the long vacation :) Will test and report.

**Comment by OctavianC** [comment]
Took a look at the code and I'm not really fond of the part where the SEF URL and the non-SEF URL are concatenated:
```php
if (preg_match('/' . preg_quote($exclusion, '/') . '/i', $this->_cache_key . ' ' . $internal_uri, $match))
```
There must be a cleaner way to do that. Why is it `preg_quoted()`? This means that regular expressions won't work, which is the point of having that feature.

Also, if matching components is a feature, why don't we add a list of all components installed and have the user select them instead of typing them in?

**Comment by andrepereiradasilva** [comment]
> Took a look at the code and I'm not really fond of the part where the SEF URL and the non-SEF URL are concatenated:
This is for performance. Why do `preg_match` operations when we can do only one?

> There must be a cleaner way to do that. Why is it preg_quoted()? This means that regular expressions won't work, which is the point of having that feature.

I don't like it too. i will change.

> Also, if matching components is a feature, why don't we add a list of all components installed and have the user select them instead of typing them in?

Yes a field for that could also be added.

**Comment by andrepereiradasilva** [comment]
Updated the code. Made some code improvements, comments and a small change to show the debug even if read from cache.

Didn't make the components select field part. Will not do it in this PR.

Please test.

**Comment by conconnl** [comment]
Can you extend your test scenario with the information how people can see if things are being cached.
Then we can let some people test this PR, improvements within the cache feature are very welcome<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8431">issues.joomla.org/joomla-cms/8431</a>.</sub>

**Comment by andrepereiradasilva** [comment]
> Can you extend your test scenario with the information how people can see if things are being cached.
> Then we can let some people test this PR, improvements within the cache feature are very welcome

you can look at the generated files in the /cache directory to see if the cache is generated or not

**Comment by euismod2336** [comment]
I have tested this item :white_check_mark: successfully on 9f0369ac3eafd22912bf3a836ccc9bebe47376f8<br /><br />Items added through the menu-item or via the advanced exclude field are not cached while others are. While site is offline no items are cached.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8431">issues.joomla.org/tracker/joomla-cms/8431</a>.</sub>

File: plugins/system/cache/cache.php [hunk 2]
```diff
@@ -32,88 +32,111 @@ public function __construct(& $subject, $config)
...
+			return;
+		}
 
-		if ($user->get('guest') && !$this->isExcluded())
+		// We need to check if user is guest again here, because auto-login plugins have not been fired before the first aid check.
+		if (JFactory::getUser()->get('guest'))
 		{
-			// We need to check again here, because auto-login plugins have not been fired before the first aid check.
+			// Saves current page in cache.
 			$this->_cache->store(null, $this->_cache_key);
```
Comment (csthomas) [inline plugins/system/cache/cache.php]: I would replace `if (JFactory::getUser()->get('guest'))` to 
`if (JFactory::getUser()->get('guest') && $this->_cache->getCaching())`
because `$this->_cache->store(...` calls code at https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/cache/controller/page.php#L129

**Comment by SamuelSchepp** [comment]
I have tested this item :white_check_mark: successfully on ad59b79d85804e0c5d05cd0f86d19972c746aa3b<br /><br />Tested:
1. Added few URIs and component names in "Exclude URLs"
1.1 /login
1.2 index.php
1.3 sample-site
2. Visited these pages
3. Edited these and other cached pages
4. Visited these pages again
Pages got cached as expacted.

Site offline testing:
Tested:
1. Set site to online
2. Enable cache plugin
3. View site
4. Set site to offline
5. Reload site

Without patch:
Cached version gets loaded.

@icampus

With patch:
Site is loaded as offline.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8431">issues.joomla.org/tracker/joomla-cms/8431</a>.</sub>

**Comment by ghost** [comment]
@euismod2336 can you please retest?

**Comment by eXsiLe95** [comment]
I have tested this item :white_check_mark: successfully on ad59b79d85804e0c5d05cd0f86d19972c746aa3b<br /><br />

## Tested:

1. Added URIs and components in "Advanced/Exclude URLS"
2. Visited this pages, made changes
3. Changed site state to offline
4. Visited the pages again
5. Changed site state to online
6. Installed bugfix
7. Visited the pages again
8. Changed site state to offline again
9. Visited the pages again

## Behaviour

### Before bugfix 

After the site was turned offline, the sites were still cached.

### After bugfix

After the site was turned offline, the site showed the preset offline message.

Tested @iCampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8431">issues.joomla.org/tracker/joomla-cms/8431</a>.</sub>

**Comment by ghost** [comment]
RTC after two successful tests.

**Comment by brianteeman** [comment]
This has merge conflicts so cant be merged as is. Not sure if @andrepereiradasilva is around to fix them

**Comment by ghost** [comment]
@andrepereiradasilva can you please resolve conflicting Files?

**Comment by mbabker** [comment]
Replaced by #21012 with merge conflicts resolved.
